### PR TITLE
feat(redaction): add custom redaction rules for project-specific identifiers

### PR DESCRIPTION
**Pull request by csansoon**
Ingest PII redaction shipped closed-world: eight built-in detectors and nothing else. A customer whose identifiers are `ACCT-` plus nine digits, or a list of 300 known account numbers, or a vendor attribute we never enumerated, had no way to redact any of it short of us shipping a detector. The ML tier that was meant to absorb that demand is not being built, so the deterministic tier has to.

This adds custom rules to the redaction policy, cascading and locking like the rest of it. Refs `specs/pii-redaction.md`.

## the three kinds

Ordered by risk, which is also the order they appear in the editor.

**`attribute_key`** replaces the whole value of a named span attribute, wherever it appears — `attr_string`, `resource_string`, and `metadata`. It masks and never removes the key, following the contract from #4333, for the same reason: a missing key is indistinguishable from one the exporter never sent. Keys match exactly or by `prefix.*`, compared with `startsWith` rather than compiled to a regex, so the kind has no backtracking surface. It is deliberately *not* gated on the metadata scope — that scope gates value scanning, and a key named explicitly is meant everywhere it appears.

**`terms`** compiles a literal alternation, so a 200-term list costs one pass per leaf. Two details are load-bearing and easy to get wrong: terms are sorted longest-first, because JS alternation is leftmost-*first*, so `ACME|ACME_CORP` otherwise matches `ACME` and leaves `_CORP` in the stored content; and word boundaries are per term, because a term with a non-word edge like `+1-555` can never match behind a blanket lookbehind.

**`pattern`** is the customer's regex, and the only kind needing the gates below.

Rules carry a label matching `^[A-Z][A-Z0-9_]{2,31}$`, which is the grammar the existing chip already shape-matches, so `[REDACTED_ACCOUNT_NUMBER]` renders like a built-in with no UI change. Built-in labels are reserved.

## how a rule is kept safe

`validateRedactionRule` answers whether a rule is **safe to run**: compiles, does not match the empty string, no backreferences, no nested unbounded quantifiers, and does not backtrack catastrophically — the last checked by scanning the source *and* by timing the pattern against generated adversarial input. One function with two callers, the editor and the write path, so the editor cannot show a verdict the write disagrees with. The gate lives in the web server functions because `@domain/spans` depends on `@domain/projects`, so validating inside the update use cases would close a cycle.

It deliberately does **not** judge over-breadth. An earlier version scored every rule against a fixed corpus of 173 strings and blocked on a hit. That rejects legitimate rules — a project whose account numbers really are long digit runs cannot express them — which is the same false-positive failure the detectors are carefully designed to avoid, built into the thing guarding them. It also asked the customer to trust a verdict about data they had never seen.

So `\d{4,}` saves. What catches it is `previewRedactionUseCase`: it samples the 50 most recent spans, runs the real policy in memory, and reports a per-label count plus the distinct changes. Two things make it readable rather than a wall — attribute maps are diffed key by key, so a change names `user.id` instead of serializing a 300-character map, and identical changes are collapsed with a span count instead of listed once per span. The summary includes labels that matched **nothing**, which is the most useful thing it can say right after someone writes a rule and is absent from the engine's counts, and identity replacements, which the engine counts separately from pattern matches and were therefore missing entirely from a pseudonymizing project's summary.

The preview never writes. That matters beyond the obvious: `toInsertRow` defaults `retention_days` to 90 while the detail read never selects it, so a read-redact-reinsert round trip would silently reset retention on every span it touched.

## cascade

Rules resolve exactly like `entities`: project list, else organization list, else empty; `locked` takes the organization list outright.

Union was considered and rejected. The shipped model is that an unlocked organization policy is a *default* a project may weaken, and `locked` is what makes it a floor. Under union a project could drop an organization *entity* but not an organization *rule*, which no UI can explain, and `locked` would shrink to meaning only "you cannot add". The ergonomic cost is paid in the UI, where the **Set by** selector already seeds a project's first override from the resolved organization values.

Rule *order* is significant and preserved: it decides which label wins an exact overlap tie. Built-ins compile first, so a validated label wins over a custom one.

## two bugs fixed on the way

Redaction sub-fields were enumerated by hand in three places, and one was live:

- `normalizeRedaction`, duplicated across both redaction use cases, compared four named fields. Any field added later was invisible to it, so a rules-only edit compared equal to the stored policy and took the no-op path: **nothing saved, no audit event, and the caller still reported success.** Without finding this the whole feature would have appeared to work while discarding every rules-only change.
- `hasRedactionField` drives `source`, and so the "following the organization default" copy and the override count. A rules-only override reported as unset.

Both are now structural rather than field lists, with `locked` excluded from the second because it gates who may change the policy rather than forming part of it. Verified by mutation: restoring the named-field comparison fails four tests, restoring the named presence check fails one.

## engine

`entities: ReadonlySet<RedactionEntity>` threaded through the walk becomes a compiled rule set of `{ label, pattern, validate? }`, and counts are keyed by label. Built-in entities compile into that shape, so the enum stays the source of truth for the eight shipped detectors while the engine no longer depends on it. Compilation happens once per distinct policy per batch, memoized by policy identity.

Two downstream surfaces needed no change, which is why this shape was chosen: the observability annotation loop already iterates `Object.entries(summary.counts)`, and the chip already shape-matches `[REDACTED_<LABEL>]` with a generic fallback for unknown labels.

Rules ride the existing flat `redaction` field on the queue payload, so it keeps its order-independent deploy property. Schema caps bound what they can add: 25 rules, 200 chars per pattern, 200 terms and 4000 chars per terms rule, 100 keys per key rule.

## scope cuts and known limits

**The public API is deferred.** Rules are dashboard-only, which matches organization redaction today. `toResponse` re-parses through `ProjectSettingsSchema`, which acts as a response allowlist, so stored rules are omitted from `GET`/`PATCH /v1/projects` until the follow-up.

An earlier version of this description claimed the T-14 pin also made the field un-writable on that path. It does not: `updateProject.execute` calls `updateProjectRedactionUseCase` directly, bypassing the pin, and that use case replaces the whole `redaction` object — so any API redaction write deleted dashboard-created rules. Fixed in 022f837d5 by preserving stored rules when the incoming setting cannot express them, covered by a use-case test and an HTTP-level one. Nothing under `packages/operations` is touched, so no manifest or SDK regeneration.

**ReDoS is bounded at write time by reading the pattern, not by running it.** Review found the original design had this backwards, and the correction is the most substantial change since: the validator's timing probe could itself be the outage. `a*a*a*b` runs in 0.4ms on 32 characters and **14 seconds on 256**, so the old 2,048-character rung blocked the event loop synchronously before any elapsed-time check could fire — reachable by any authenticated session.

No probe length fixes that, because one long enough to expose polynomial blowup is long enough to wedge the process. So the source scanner now carries the defence and the probe is only a backstop at lengths that cannot hurt. It refuses nested unbounded quantifiers, adjacent unbounded quantifiers over atoms that can match the same character (`a*a*`, `\w+\d+`), unboundedly repeated alternations whose branches can start alike (`(a|aa)+`), and backreferences. Overlap is measured by asking the engine rather than assumed, so `[A-Z]+\d*`, `\d+-\d+` and `(?:\d|-)+` stay allowed — and a test pins that none of the 26 shipped detector patterns trips the new gate, which is the only corpus of real PII shapes available to calibrate against.

What is still not closed: a single `matchAll` is not interruptible, so the per-leaf batch deadline bounds a span carrying hundreds of slow leaves but not one catastrophic leaf. The gates are static heuristics over a hand-written scanner, not a parser, so a shape neither the scanner nor the probe recognises would still reach ingest. A validator in a killable worker is the complete answer and RE2 the durable one. Each stored pattern records the `validatorVersion` that admitted it, so a tightened validator can flag survivors instead of trusting them.

**Custom rules do not reach span names, `model`, `serviceName`, or other identifier and enum columns.** People do put account ids in span names and will ask. Those are grouping keys, so redacting them breaks every aggregate; if we ever allow it, it is a separate opt-in scope rather than a rule property. Said in the docs so support does not field it.

## validation

1901 tests in `@domain/spans`, 874 in `@app/web`, 215 in `@domain/shared`, 46 in `@domain/organizations`, 27 in `@domain/projects`, plus 42 in `@platform/db-clickhouse`. `pnpm typecheck` clean across all 90 packages; `knip` and `format` clean.

Worth knowing about the tests specifically:

- Integration coverage runs the real `transformOtlpToSpans`, because a hand-built span cannot catch the leak that matters: every string attribute is copied verbatim into `attr_string`, so a rule reaching only the parsed columns would leave a full plaintext copy in the same row.
- `listRecentDetailsByProjectId` had no test despite existing since the v2 rewrite with a doc comment describing exactly this use. Added two, covering recency ordering and the `LIMIT 1 BY` dedupe.
- `SAFE_CORPUS` survives the corpus removal as a regression guard asserting no built-in detector matches any of it — a guard the per-entity negative vectors cannot give, since each only ever runs its own detector.

**Not yet done:** the end-to-end pass against a running stack, and a re-run of the §4.7 benchmark with custom rules active. The end-to-end run matters more than usual here, because after the corpus removal the preview is the only thing between a greedy pattern and a project's tool output.

## Summary by CodeRabbit

* **New Features**
  * Added custom redaction rules for exact terms, attribute keys, and regular expressions.
  * Added controls to create, edit, enable, disable, reorder, and remove rules.
  * Added a read-only preview with affected spans, matches, and before/after examples.
  * Added identity-handling choices for retaining or pseudonymizing identifiers.
* **Bug Fixes**
  * Invalid or unsafe rules are rejected with clear explanations.
  * Redaction-only policy changes are correctly saved and recorded.
* **Documentation**
  * Expanded guidance for custom rules, inheritance, validation, and preview behavior.

### CODE REVIEW

File: packages/domain/shared/src/settings.ts [hunk 1]
```diff
@@ -47,6 +146,7 @@ export const redactionSettingSchema = z.object({
   entities: z.array(redactionEntitySchema).optional(),
   scopes: redactionScopesSettingSchema.optional(),
   identities: redactionIdentityHandlingSchema.optional(),
+  rules: z.array(redactionRuleSchema).max(REDACTION_MAX_RULES).optional(),
```
[the remark this answers is not in the document]
Comment (csansoon) [reply packages/domain/shared/src/settings.ts:153]: Confirmed and fixed in 022f837d5.

The mechanism is slightly different from the description, and worse: `updateProject.execute` does not go through the T-14 pin at all — it calls `updateProjectRedactionUseCase` directly with `body.settings.redaction`, and that use case replaces the whole `redaction` object. So any API redaction write dropped stored rules, and the PR description's claim that the field was merely "inert on that path" was wrong.

Fixed by preserving stored `rules` when the incoming setting cannot express them (`withPreservedRules`, in both the project and organization use cases). A write that does carry `rules` still replaces them, so the dashboard keeps its delete path. Covered by three use-case tests plus an HTTP-level one in `apps/api/src/routes/projects.test.ts`, mutation-verified.

File: packages/domain/spans/src/redaction/redact-span.ts [hunk 2]
```diff
@@ -73,17 +77,22 @@ export function redactSpanDetail(
     return outcome.value
   }
 
-  const redactedAttrString = take(redactStringMap(substitute(span.attrString), entities))
-  const resourceString = take(redactStringMap(substitute(span.resourceString), entities))
+  // A key rule masks the whole value wherever the key appears. Unlike the value pass it is not
+  // gated on the metadata scope: a key named explicitly is meant everywhere, and masking one
+  // cannot produce a false positive.
+  const maskedAttrString = take(maskKeyedValues(substitute(span.attrString), ruleSet))
+  const maskedResourceString = take(maskKeyedValues(substitute(span.resourceString), ruleSet))
```
[the remark this answers is not in the document]
Comment (csansoon) [reply packages/domain/spans/src/redaction/redact-span.ts:95]: Confirmed and fixed in dc630a99c.

A key rule names an attribute, not a column type, so it now reaches `attrInt`, `attrFloat` and `attrBool` via `maskKeyedScalars`, relocating the matched key into `attrString` as the placeholder (a `Map(String, Int64)` cannot hold one). Value scanning stays out of the boolean map, since no detector matches `true`.

The test asserts the pre-fix state as well, so it cannot pass vacuously.

File: packages/domain/spans/src/redaction/validate-rule.ts [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+
+        return slowest
+      }
+    }
+  }
+
+  for (const length of POLYNOMIAL_PROBE_LENGTHS) {
+    for (const character of alphabet) {
+      const elapsed = timeAgainst(pattern, character.repeat(length))
+      slowest = Math.max(slowest, elapsed)
```
[the remark this answers is not in the document]
Comment (csansoon) [reply packages/domain/spans/src/redaction/validate-rule.ts]: Confirmed, and it measured worse than described. Fixed in 0e52c0016.

`a*a*a*b` takes **14 seconds at 256 characters** and 0.4ms at 32, so the old ladder jumping 32 to 2,048 would have blocked the event loop for minutes.

Your last sentence is the important one: a timeout measured after `RegExp.test` returns cannot work, and there is no probe length that reveals polynomial blowup safely — one long enough to expose it is long enough to be the outage. Rather than lengthen or interrupt the probe, the source scanner now carries the defence and the probe is only a backstop:

- **adjacent** unbounded quantifiers over atoms that can match the same character (`a*a*`, `\w+\d+`) — the polynomial case
- unboundedly repeated **alternations whose branches can start alike** (`(a|aa)+`) — the one exponential shape that stays under budget at every length the probe can safely reach

Overlap is measured by asking the engine, not assumed, so `[A-Z]+\d*`, `\d+-\d+` and `(?:\d|-)+` stay allowed. A test pins that none of the 26 shipped detector patterns trips the new gate, and another pins that each catastrophic shape is refused in under 250ms rather than by running to completion.

The ladder is also now stepped by 4 rather than 8: exponential cost doubles per character, so a step of 8 let a rung that passed at the budget be followed by one costing 256x as much.

A worker with a hard kill is still the only complete answer, and the residual risk is stated in the PR description. This closes the reachable case.

File: apps/web/src/domains/projects/projects.functions.ts [hunk 4]
```diff
@@ -218,6 +242,60 @@ export const updateProjectRedaction = createServerFn({ method: "POST" })
...
+    // Branded org id from `resolveOrgScope`, never the raw one: ClickHouse has no RLS, so this is
+    // the only thing keeping the read inside the tenant.
+    const orgId = await resolveOrgScope(context)
+
+    return Effect.runPromise(
+      previewRedactionUseCase({
+        organizationId: orgId,
+        projectId: ProjectId(data.projectId),
+        policy: resolveRedactionPolicy({ organization: null, project: { redaction: data.redaction } }),
+        sampleSize: PREVIEW_SAMPLE_SIZE,
```
[the remark this answers is not in the document]
Comment (csansoon) [reply apps/web/src/domains/projects/projects.functions.ts:318]: Confirmed and fixed in 5415acc2a.

The preview was admin-gated but fed `data.redaction` straight into the engine, and the inputs are worse than the write path's: stored tool outputs are far longer than anything the validator probes with. It now runs the same `rejectInvalidRedactionRules` gate before reaching the use case.

Worth noting `(a+)+$` specifically was already refused by the source scanner, so the reachable-in-practice case was a pattern the scanner missed. Both halves are now closed: the scanner is stricter (see the thread on `validate-rule.ts`) and the preview no longer accepts what the save rejects.

File: apps/web/src/domains/organizations/organizations.functions.ts [hunk 5]
```diff
@@ -30,6 +30,7 @@ import { createServerFn } from "@tanstack/react-start"
 import { getRequestHeaders } from "@tanstack/react-start/server"
 import { Effect, Layer } from "effect"
 import { z } from "zod"
+import { rejectInvalidRedactionRules } from "../../lib/redaction-rules.ts"
```
[the remark this answers is not in the document]
Comment (csansoon) [reply apps/web/src/domains/organizations/organizations.functions.ts]: Fixed in 5415acc2a, at both sites. Extracted as `rejectionMessage` in `lib/redaction-rules.ts` so the two callers cannot drift, and it quotes the label as you suggested: `Rule "ACCOUNT_NUMBER": two rules share the id r1`.

File: apps/web/src/domains/projects/projects.functions.ts [hunk 6]
```diff
@@ -218,6 +242,60 @@ export const updateProjectRedaction = createServerFn({ method: "POST" })
...
+ * verdict the save disagrees with. It lives on the server because the corpus and the engine are
+ * in `@domain/spans`, which has no business in the browser bundle.
+ */
+export const validateRedactionRuleDraft = createServerFn({ method: "POST" })
+  .inputValidator(z.object({ rule: redactionRuleSchema }))
+  .handler(async ({ data }): Promise<RuleValidation> => {
+    await requireSession()
+
+    return validateRedactionRule(data.rule)
+  })
```
[the remark this answers is not in the document]
Comment (csansoon) [reply apps/web/src/domains/projects/projects.functions.ts:273]: Fixed in 5415acc2a. Gated on `memberships.isAdmin` like its siblings.

Your second reading was the right one to check: the endpoint touches no project data, which is why it was missed. But it does run a caller-supplied pattern on the server, and only the people who can save a rule have any reason to ask what one would do, so consistency is the correct call regardless. The cost of that pattern is also now bounded — see the `validate-rule.ts` thread, where the probe could previously block for seconds.

File: apps/web/src/domains/projects/redaction-rule-drafts.ts [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+ * Throws rather than falling back to an empty list. An empty list is a valid policy that deletes
+ * every rule the customer had, so swallowing a decode failure here would quietly destroy their
+ * configuration; failing the apply surfaces it instead.
+ */
+export const decodeRules = (encoded: string): RedactionRule[] => {
+  const parsed = encodedRulesSchema.safeParse(JSON.parse(encoded))
+  if (!parsed.success) throw new Error("Could not read the redaction rules on this page. Reload and try again.")
+
+  return parsed.data
+}
```
[the remark this answers is not in the document]
Comment (csansoon) [reply apps/web/src/domains/projects/redaction-rule-drafts.ts:79]: Fixed in 5415acc2a, and you were right that the test was the more useful half of this. `JSON.parse` is wrapped so both unparseable and well-formed-but-wrong input produce the same actionable message, and the test now asserts on that message across four inputs (`""` and `"null"` added) rather than merely that something throws.

File: apps/web/src/routes/_authenticated/projects/$projectSlug/settings/-components/redaction-rule-sheet.tsx [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+    }
+  }, [open, ready, draft])
+
+  const changeKind = (kind: RedactionRuleKind) => {
+    // Kind-specific fields cannot carry over, so keep only what every kind shares.
+    setDraft({ ...newRuleDraft(kind), id: draft.id, label: draft.label })
+  }
+
+  const labelError = labelIssue(draft.label)
+  const canSave = ready && labelError === undefined && validation?.ok === true
```
[the remark this answers is not in the document]
Comment (csansoon) [reply apps/web/src/routes/_authenticated/projects/$projectSlug/settings/-components/redaction-rule-sheet.tsx]: Confirmed and fixed in 5415acc2a, with both changes you proposed: `setValidation(null)` when the draft changes, and `isValidating` in `canSave`.

The consequence you identified in passing is the sharpest part — Save stamps `validatorVersion` off `validation` for pattern rules, so the stale verdict was not just a disabled-state mismatch, it attributed the old draft's approval to a pattern nobody had checked.

Added `redaction-rule-sheet.test.tsx` covering it, and mutation-verified: reverting the two lines fails the race test and leaves the other two passing.

File: packages/domain/spans/src/index.ts [hunk 9]
```diff
@@ -267,6 +267,8 @@ export type {
   UserUsageSlice,
 } from "./ports/user-analytics-repository.ts"
 export { isUserSortField, USER_SORT_FIELDS, UserAnalyticsRepository } from "./ports/user-analytics-repository.ts"
+export type { RuleValidation, RuleValidationIssue } from "./redaction/validate-rule.ts"
+export { REDACTION_VALIDATOR_VERSION, validateRedactionRule } from "./redaction/validate-rule.ts"
```
[the remark this answers is not in the document]
Comment (csansoon) [reply packages/domain/spans/src/index.ts:271]: Declining this one — the premise does not hold, and the change would be actively harmful.

`validateRedactionRule` is never imported into a browser module. Its only two importers are server-only: `projects.functions.ts` (inside a `createServerFn` handler) and `lib/redaction-rules.ts`, which is itself consumed only by those server functions. The component imports `RuleValidation` with `import type`, which is erased at compile time and never resolved at runtime, and it reaches the validator over the network via the `validateRedactionRuleDraft` server function. `pnpm --filter @app/web typecheck` passes, so nothing is resolving in a broken way.

Adding `validateRedactionRule` to `browser.ts` would pull the redaction engine — every detector pattern plus the safe corpus — into the client bundle, which is exactly what the doc comment on that server function says must not happen. The absence from `browser.ts` is the design, not an oversight.

File: packages/domain/spans/src/redaction/safe-corpus.test.ts [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+      matched: entry.slice(match.start, match.end),
+    }))
+
+    expect(matches).toEqual([])
+  })
+
+  it("covers every entity that ships on by default", () => {
+    // Not a coverage metric, just a floor: a corpus that shrank to nothing would still pass above.
+    expect(SAFE_CORPUS.length).toBeGreaterThan(150)
+  })
```
[the remark this answers is not in the document]
Comment (csansoon) [reply packages/domain/spans/src/redaction/safe-corpus.test.ts]: Fixed in 0e52c0016, using your title verbatim. Also updated the neighbouring duplicate test, whose rationale ("would double-count a rule's corpus hits") was stale for the same reason as the comment you flagged in `validate-rule.test.ts` — corpus scoring is gone, so the reason to forbid duplicates is now that they pad the floor without widening what it guards.

File: packages/domain/spans/src/redaction/validate-rule.test.ts [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+    "urn:acme:customer:[0-9a-f]{8}",
+  ])("accepts the legitimate pattern %s", (pattern) => {
+    expect(validateRedactionRule(patternRule(pattern))).toMatchObject({ ok: true, errors: [] })
+  })
+
+  /**
+   * Both bounds have to be unbounded for star height to blow up. Treating every quantified group
+   * as suspect would reject ordinary patterns, so these are checked against the quantifier gate
+   * specifically — they are over-broad for other reasons and the corpus gate rejects them on that.
+   */
```
[the remark this answers is not in the document]
Comment (csansoon) [reply packages/domain/spans/src/redaction/validate-rule.test.ts:37]: Fixed in 0e52c0016 with your wording. Two more stale corpus references turned up alongside it and are gone too: `lib/redaction-rules.ts` still said the ingest path skipped "a corpus scan and a timing probe", and `safe-corpus.test.ts` justified its duplicate check by "corpus hits".

File: packages/domain/spans/src/redaction/validate-rule.ts [hunk 12]
```diff
@@ -0,0 +1,300 @@
...
+    if (character === "(") {
+      groups.push({ unbounded: false })
+      // Skip the group-type prefix so `?` in `(?:`, `(?=`, `(?<name>` is not read as a quantifier.
+      index += source.startsWith("(?", index) ? 2 : 1
+      if (source[index] === "<") {
+        const close = source.indexOf(">", index)
+        index = close === -1 ? index + 1 : close + 1
+      }
+      continue
+    }
```
[the remark this answers is not in the document]
Comment (csansoon) [reply packages/domain/spans/src/redaction/validate-rule.ts:245]: Confirmed and fixed in 0e52c0016, with your exact example as the regression test.

Verified the mechanism first: for `(?<![0-9])(\d+)+>` the cursor jumped to the final `>` and swallowed `(\d+)+`, so the `catastrophic_backtracking` I saw in that case came from the *timing probe*, not the scanner — which matters, because the probe turned out to be unsafe for exactly this reason (see the `probePattern` thread). The scanner now carries the defence, so a hole in it was no longer just a missed report.

Went slightly further than the diff: prefix handling moved into `skipGroupPrefix`, which also consumes the `:` in `(?:`. That `:` was being read as a literal atom — harmless before, but it corrupted the new alternation-ambiguity check by standing in for the group's real first atom.

File: packages/domain/spans/src/redaction/validate-rule.ts [hunk 13]
```diff
@@ -0,0 +1,300 @@
...
+          message: `took ${elapsed.toFixed(0)}ms on ${length} characters, which is too slow for the ingest path`,
+        })
+
+        return slowest
+      }
+    }
+  }
+
+  return slowest
+}
```
[the remark this answers is not in the document]
Comment (csansoon) [reply packages/domain/spans/src/redaction/validate-rule.ts:447]: Adopted in 0e52c0016, including the early break from your diff — which turned out to be load-bearing rather than an optimisation. Without it a slow pattern pays its cost three times: my own timing test caught `(a|aa)+$` going from ~170ms to 500ms. Best-of-N is there to rescue a marginal rule from noise, and a pattern over budget by a wide margin is not marginal.

The wider point stands and the budget moved for it, though in the opposite direction from what you might expect: the 8,192-character probe is gone entirely (a probe that long cannot be made safe — see the `probePattern` thread), and with the ladder capped at 24 characters the budget is now 8ms. That is still four orders of magnitude of headroom, since a linear pattern scans 24 characters in microseconds, so noise has less room to decide a verdict than before, not more.

File: packages/domain/spans/src/redaction/validate-rule.test.ts [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+    expect(validation.slowestProbeMs).toBeGreaterThanOrEqual(0)
+  })
+
+  /**
+   * The syntactic gate is a heuristic, so the probe is the backstop that does not depend on
+   * recognising a shape. This pattern nests its repetition across an alternation rather than a
+   * group boundary, which is exactly the case a star-height scan is weakest on.
+   */
+  it("catches catastrophic backtracking the source scan alone would miss", () => {
+    const validation = validateRedactionRule(patternRule("(?:a|aa)+$"))
```
[the remark this answers is not in the document]
Comment (csansoon) [reply packages/domain/spans/src/redaction/validate-rule.test.ts]: These two are deliberate test fixtures, not production patterns. They are the inputs asserting that `validateRedactionRule` **rejects** catastrophic regexes — the alert is firing on the thing being guarded against.

They are never used to redact anything: the validator compiles them, refuses them, and the test asserts the refusal. As of 0e52c0016 both are now rejected by the source scanner without being executed at all, so the dataflow CodeQL is tracing no longer reaches a run on attacker-controlled input.

There is no file-based way to mark this in the repo — scanning is GitHub default setup, with no `codeql-config.yml` — so these want dismissing as "used in tests" in the Security tab. Leaving that call to a human rather than dismissing security alerts from an agent.

File: packages/domain/spans/src/redaction/validate-rule.test.ts [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+
+    expect(validation.ok).toBe(false)
+    expect(validation.errors.map((issue) => issue.code)).toContain("catastrophic_backtracking")
+  })
+
+  it("finishes quickly even when rejecting a catastrophic pattern", () => {
+    const started = performance.now()
+    validateRedactionRule(patternRule("(a+)+$"))
```
[the remark this answers is not in the document]
Comment (csansoon) [reply packages/domain/spans/src/redaction/validate-rule.test.ts:193]: These two are deliberate test fixtures, not production patterns. They are the inputs asserting that `validateRedactionRule` **rejects** catastrophic regexes — the alert is firing on the thing being guarded against.

They are never used to redact anything: the validator compiles them, refuses them, and the test asserts the refusal. As of 0e52c0016 both are now rejected by the source scanner without being executed at all, so the dataflow CodeQL is tracing no longer reaches a run on attacker-controlled input.

There is no file-based way to mark this in the repo — scanning is GitHub default setup, with no `codeql-config.yml` — so these want dismissing as "used in tests" in the Security tab. Leaving that call to a human rather than dismissing security alerts from an agent.

File: apps/web/src/routes/_authenticated/projects/$projectSlug/settings/-components/redaction-rule-sheet.tsx [hunk 16]
```diff
@@ -0,0 +1,230 @@
...
+      validateRedactionRuleDraft({ data: { rule: draft } })
+        .then((next) => {
+          if (!cancelled) setValidation(next)
+        })
+        .catch(() => {
+          if (!cancelled) setValidation(null)
+        })
+        .finally(() => {
+          if (!cancelled) setIsValidating(false)
+        })
```
[the remark this answers is not in the document]
Comment (csansoon) [reply apps/web/src/routes/_authenticated/projects/$projectSlug/settings/-components/redaction-rule-sheet.tsx:72]: Good catch, and this one was made worse by the fix in the neighbouring thread: clearing the stale verdict turned "shows a stale verdict after a failure" into "shows Checking forever". Fixed in 9121764a6.

Took your framing rather than adding a third flag — `null` standing for two states was the actual defect, so the nullable verdict plus `isValidating` is now a single `Verdict` (`checking` | `checked` | `unavailable`). That makes the failure representable, the panel exhaustive over it, and `canSave` a single check instead of three that had to agree. The failure case now says the check could not run and that Save stays disabled until it does. Covered by a test asserting the message appears and "Checking the rule…" does not.

File: apps/web/src/routes/_authenticated/projects/$projectSlug/settings/privacy.tsx [hunk 17]
```diff
@@ -248,6 +254,10 @@ function ProjectPrivacySettingsPage() {
             onChange={(key, next) => setField(key, next)}
           />
         </ScopedSetting>
+
+        {view.mode === "enforce" ? (
+          <RedactionPreview projectId={currentProject.id} disabled={!canEditProject} setting={toSetting(view)} />
+        ) : null}
```
[the remark this answers is not in the document]
Comment (csansoon) [reply apps/web/src/routes/_authenticated/projects/$projectSlug/settings/privacy.tsx]: Both fixed in 9121764a6.

On the throw: `toSettingOrNull` is used on the render path while `apply` keeps the throwing version, since the asymmetry is deliberate — a decode failure must never be saved as an empty rule list, but it should only cost the preview, not the page.

On the stale result: fixed by keying `RedactionPreview` on the draft rather than resetting inside it. Every draft field is a primitive by the overlay's own design (that is why `rules` is an encoded string), so `JSON.stringify(view)` is a stable identity and React discards the old result on any policy edit — no extra effect, and it cannot drift from whatever fields the card gains later.

File: packages/domain/spans/src/redaction/validate-rule.ts [hunk 18]
```diff
@@ -0,0 +1,476 @@
...
+  "é",
+]
+
+const atomMatcher = (atom: string): RegExp | null => {
+  try {
+    return new RegExp(`^(?:${atom})$`, "u")
+  } catch {
+    return null
+  }
+}
```
[the remark this answers is not in the document]
Comment (csansoon) [reply packages/domain/spans/src/redaction/validate-rule.ts]: Confirmed and fixed in 69890406d. Best finding of the round — this was a real hole in the gate I had just argued carries the whole ReDoS defence.

Measured before fixing, and the three cases split two ways:

- **`[a-z]+[A-Z]+x` with `ignoreCase` passed everything** — empty error list, `ok: true`. A polynomial pattern reaching ingest, which is exactly the class the scanner has to catch because the probe cannot see it at safe lengths.
- `(a|A)+$` and `(?:x|X)+y` were blocked, but only by the probe as `catastrophic_backtracking`, and only because they happen to exceed the budget by 24 characters. `(a|A)+$` under `i` did not finish a 256-character input at all in a standalone measurement.

So the effect was worse than a mislabelled error on two of them and a straight escape on the third. Threaded `ignoreCase` through `validatePatternSource`, `hasOverlappingBranches` and `atomsOverlap`, with `atomMatcher` compiling `ui` when set.

Tests assert each shape is now refused from the source in under 100ms, and that all three stay accepted when case is significant — so the flag decides rather than the shape, and the fix cannot have been a blanket tightening. Mutation-verified by reverting the flag.

`dotAll` deliberately not threaded: `.` is already treated as overlapping everything, so `s` cannot change an overlap answer.

**Comment by csansoon** [comment]
Picked up the two nitpicks from the latest reviews as well.

**`withPreservedRules` duplicated across the two use cases** — fixed in 8f01e5715. Agreed on the reasoning: it encodes a compliance rule (omission preserves, empty array clears), so two copies differing only in a type name is the kind of duplication that ends with one layer deleting customer rules. Now one generic in `@domain/shared`, next to the redaction types both use cases already import.

Worth recording what that refactor turned up. Mutation-testing the shared body failed a project test and **no organization test at all** — the org path had never been pinned, so my earlier claim of mirrored coverage was wrong. Added the missing preservation test on that side plus its explicit-empty counterpart; both layers now fail if the rule regresses.

**Multi-line comments in the `.tsx` files** — fixed in 7625ea27d, all three plus two more of the same shape I had added after that review. Each keeps the constraint it recorded, just in one line.

Verification after all four commits: 1907 tests in `@domain/spans`, 874 in `@app/web`, 215 in `@domain/shared`, 48 in `@domain/organizations`, 27 in `@domain/projects`. `pnpm typecheck` clean across all 90 packages, `knip` and `format` clean.
